internal/pkg/cli/svc_deploy.go

```diff
@@ -309,26 +314,67 @@ func (o *deploySvcOpts) pushToECRRepo() error {
 	return o.docker.Push(uri, o.ImageTag)
 }
 
-func (o *deploySvcOpts) getDockerfilePath() (string, error) {
-	type dfPath interface {
+type dockerParams struct {
+	context    string
+	dockerfile string
+	args       map[string]string
+}
+
+// toDockerBuildArguments fills in the fields contained in the image/build tag in the manifest.
+func (d *dockerParams) toDockerBuildArguments() docker.BuildArguments {
+	return docker.BuildArguments{
+		Context:    d.context,
+		Dockerfile: d.dockerfile,
+		Args:       d.args,
+	}
+}
+
+func (o *deploySvcOpts) getDockerfile() (*dockerParams, error) {
+	type dfContext interface {
+		DockerfileContext() string
 		DockerfilePath() string
+		DockerArgs() map[string]string
 	}
 
 	manifestBytes, err := o.ws.ReadServiceManifest(o.Name)
 	if err != nil {
-		return "", fmt.Errorf("read manifest file %s: %w", o.Name, err)
+		return nil, fmt.Errorf("read manifest file %s: %w", o.Name, err)
 	}
-
-	svc, err := manifest.UnmarshalService(manifestBytes)
+	svc, err := o.unmarshal(manifestBytes)
 	if err != nil {
-		return "", fmt.Errorf("unmarshal svc manifest: %w", err)
+		return nil, fmt.Errorf("unmarshal service %s manifest: %w", o.Name, err)
 	}
-
-	mf, ok := svc.(dfPath)
+	mf, ok := svc.(dfContext)
 	if !ok {
-		return "", fmt.Errorf("service %s does not have a dockerfile path", o.Name)
+		return nil, fmt.Errorf("service %s does not have required methods DockerArgs, DockerfilePath, and DockerfileContext", o.Name)
 	}
-	return mf.DockerfilePath(), nil
+	copilotDir, err := o.ws.CopilotDirPath()
+	if err != nil {
+		return nil, fmt.Errorf("get copilot directory: %w", err)
+	}
+	wsRoot := filepath.Dir(copilotDir)
+
+	dfPath := filepath.Join(wsRoot, mf.DockerfilePath())
+
+	dfArgs := mf.DockerArgs()
+
+	ctx := mf.DockerfileContext()
+	// Nonempty context field means we should join with the ws root and return that
+	if ctx != "" {
+		return &dockerParams{
+			dockerfile: dfPath,
+			context:    filepath.Join(wsRoot, mf.DockerfileContext()),
+			args:       dfArgs,
+		}, nil
+	}
+	// Default to directory of dockerfile.
+
+	dfDir := filepath.Dir(dfPath)
+	return &dockerParams{
```

Comment:
Should we return `.BuildString` instead here?

Reply:
I don't necessarily think so, since BuildString is more of a convenience class to allow us to unmarshal YAML of variable type. I'd like to enforce passing only docker.BuildArguments instances to docker.Build().

Reply:
Sounds good!